java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -0,0 +1,187 @@
+   */
+  private List<Symbol> definedInState = Lists.newArrayList();
+  private Map<Value, State> stateOfValue = Maps.newHashMap();
+
+  public ExecutionState(ExecutionState executionState) {
+    this.parent = executionState;
+    this.check = executionState.check;
+    this.reachableValues = HashMultimap.create(executionState.reachableValues);
+    this.unreachableValues = HashMultimap.create(executionState.unreachableValues);
+  }
+
+  public ExecutionState(IssuableSubscriptionVisitor check) {
+    this.check = check;
+  }
+
+  public void defineSymbol(Symbol symbol) {
+    definedInState.add(symbol);
+  }
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Symbol symbol : executionState.reachableValues.keys()) {
+      this.reachableValues.putAll(symbol, executionState.reachableValues.get(symbol));
+    }
+    for (Symbol symbol : executionState.unreachableValues.keys()) {
+      this.unreachableValues.putAll(symbol, executionState.unreachableValues.get(symbol));
+    }
+
+    for (Symbol symbol : unreachableValues.keys()) {
+      //cleanup after merge of reachable/unreachable values
+      for (Value value : unreachableValues.get(symbol)) {
+        reachableValues.remove(symbol, value);
+      }
+    }
+    //Merge states of values
+    for (Map.Entry<Value, State> valueStateEntry : executionState.stateOfValue.entrySet()) {
+      Value value = valueStateEntry.getKey();
+      State state = valueStateEntry.getValue();
+      State valueState = getStateOfValue(value);
+      if(valueState == null) {
+        valueState = state;
+      } else {
+        valueState = valueState.merge(state);
+      }
+      this.stateOfValue.put(value, valueState);
+    }
+    return this;
+  }
+
+  public ExecutionState overrideBy(ExecutionState executionState) {
+    this.unreachableValues.putAll(executionState.unreachableValues);
+    this.reachableValues = executionState.reachableValues;
+    this.stateOfValue.putAll(executionState.stateOfValue);
+    return this;
+  }
+
+  public ExecutionState restoreParent() {
+    if (parent != null) {
+      insertIssues();
+      return parent.merge(this);
+    }
+    return this;
+  }
+
+  public void insertIssues() {
+    for (Tree tree : getIssuableTrees()) {
+      check.addIssue(tree, "");
+    }
+  }
+
+  private Set<Tree> getIssuableTrees() {
+    Set<Tree> results = Sets.newHashSet();
+    for (Symbol symbol : definedInState) {
+      for (Value value : unreachableValues.get(symbol)) {
+        State state = stateOfValue.get(value);
+        if (state.shouldRaiseIssue()) {
+          results.addAll(state.reportingTrees());
+        }
+      }
+      for (Value value : reachableValues.get(symbol)) {
+        State state = stateOfValue.get(value);
+        if (state.shouldRaiseIssue()) {
+          results.addAll(state.reportingTrees());
+        }
+      }
+    }
+    return results;
+  }
+
+  //FIXME : Hideous hack for closeable to get "Ignored" variables
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Take the required action to fix the issue indicated by this comment.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)